### PR TITLE
[feature](mtmv)(3)Implementing multi table materialized views

### PR DESCRIPTION
**Pull request by zddr**
## Proposed changes

Issue Number: close #xxx

Implement mtmv

Introduction to Main Classes：
- MTMVService：MTMV services for other modules to call
- MTMVHookService：All operations that affect the MTMV
  - MTMVJobManager：All operations that affect the MTMV job
  - MTMVCacheManager：All operations that affect the MTMV Cache
- MTMVTask&MTMVJob：Inherit from job framework

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/job/extensions/mtmv/MTMVTask.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    private MTMVCache cache;
+
+    public MTMVTask(String dbName, long mtmvId) {
+        this.dbName = dbName;
+        this.mtmvId = mtmvId;
+    }
+
+    @Override
+    public void run() throws JobException {
+        ConnectContext ctx = createContext();
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/job/extensions/mtmv/MTMVTask.java:70]: you create a ctx here, and create another ctx in `generateMTMVCache() -> getPlanBySql()` again?
And in ctx, can we use the user identity who created this job?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java [hunk 2]
```diff
@@ -5741,4 +5759,28 @@ public ColumnIdFlushDaemon getColumnIdFlusher() {
     public StatisticsAutoCollector getStatisticsAutoCollector() {
         return statisticsAutoCollector;
     }
+
+    public void alterMTMVRefreshInfo(AlterMTMVRefreshInfo info) throws UserException {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Env.java:5763]: The following 4 methods can be moved to `Alter` class?

File: fe/fe-core/src/main/java/org/apache/doris/alter/Alter.java [hunk 3]
```diff
@@ -875,4 +872,36 @@ public AlterHandler getMaterializedViewHandler() {
...
+
+    public void processAlterMTMV(AlterMTMV alterMTMV, boolean isReplay) {
+        TableNameInfo tbl = alterMTMV.getMvName();
+        MTMV mtmv = null;
+        try {
+            Database db = Env.getCurrentInternalCatalog().getDbOrDdlException(tbl.getDb());
+            mtmv = (MTMV) db.getTableOrMetaException(tbl.getTbl(), TableType.MATERIALIZED_VIEW);
+
+            mtmv.writeLock();
+            if (alterMTMV.getRefreshInfo() != null) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/alter/Alter.java:884]: Better use a enum `type` to specify what kind of this alter is?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/MTMV.java [hunk 4]
```diff
@@ -17,16 +17,185 @@
...
+    private ReentrantReadWriteLock mvRwLock;
+
+    @SerializedName("ri")
+    private MTMVRefreshInfo refreshInfo;
+    @SerializedName("qs")
+    private String querySql;
+    @SerializedName("s")
+    private MTMVStatus status;
+    @SerializedName("ei")
+    private EnvInfo envInfo;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/MTMV.java:57]: `EnvInfo` only contains ctl and database, the class name is not suitable?

File: fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVCache.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+import com.google.gson.annotations.SerializedName;
+
+import java.util.Set;
+
+public class MTMVCache {
+    @SerializedName("bt")
+    private Set<BaseTableInfo> baseTables;
+    @SerializedName("bv")
+    private Set<BaseTableInfo> baseViews;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVCache.java:28]: What is diff between `baseTables` and `baseViews`?

File: fe/fe-core/src/main/java/org/apache/doris/job/base/AbstractJob.java [hunk 6]
```diff
@@ -132,23 +153,16 @@ public void updateJobStatus(JobStatus newJobStatus) {
...
-            case MTMV:
-                // return MTMVJob.readFields(in);
-                break;
-            default:
-                throw new IllegalArgumentException("unknown job type");
-        }
-        throw new IllegalArgumentException("unknown job type");
+        String jsonJob = Text.readString(in);
+        AbstractJob<?> job = GsonUtils.GSON.fromJson(jsonJob, AbstractJob.class);
+        job.setRunningTasks(new ArrayList<>());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/job/base/AbstractJob.java:158]: Why  need to set running tasks to empty?
The running tasks is not persisted.

File: fe/fe-core/src/main/java/org/apache/doris/job/extensions/insert/InsertTask.java [hunk 7]
```diff
@@ -34,44 +53,137 @@ public class InsertTask extends AbstractTask {
...
     @Override
-    public void before() {
+    public void before() throws JobException {
+        if (isCanceled.get()) {
+            throw new JobException("Export executor has been canceled, task id: {}", getTaskId());
+        }
+        ctx = new ConnectContext();
+        ctx.setEnv(Env.getCurrentEnv());
+        ctx.setCluster(SystemInfoService.DEFAULT_CLUSTER);
+        ctx.setQualifiedUser(Auth.ADMIN_USER);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/job/extensions/insert/InsertTask.java:83]: Use user who create this job?

File: fe/fe-core/src/main/java/org/apache/doris/job/extensions/mtmv/MTMVTask.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    public static final Long MAX_HISTORY_TASKS_NUM = 100L;
+
+    @SerializedName(value = "dn")
+    private String dbName;
+    @SerializedName(value = "mi")
+    private long mtmvId;
+    @SerializedName("sql")
+    private String sql;
+
+    private MTMV mtmv;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/job/extensions/mtmv/MTMVTask.java:60]: These can be set to null after task finished

File: fe/fe-core/src/main/java/org/apache/doris/job/extensions/mtmv/MTMVJob.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+    }
+
+    @Override
+    public List<MTMVTask> createTasks(TaskType taskType) {
+        MTMVTask task = new MTMVTask(dbName, mtmvId);
+        task.setTaskType(taskType);
+        ArrayList<MTMVTask> tasks = new ArrayList<>();
+        tasks.add(task);
+        super.initTasks(tasks);
```
[the remark this answers is not in the document]
Comment (CalvinKirs) [reply fe/fe-core/src/main/java/org/apache/doris/job/extensions/mtmv/MTMVJob.java:93]: This belongs to the task's own creation logic, and there may be other operations that need to be processed on the data that has been initialized. At the same time, I am not sure if the implementation class will have custom init logic (sure,this is usually not necessary).

File: fe/fe-core/src/main/java/org/apache/doris/job/base/AbstractJob.java (hunk 6 above)
Comment (CalvinKirs) [reply fe/fe-core/src/main/java/org/apache/doris/job/base/AbstractJob.java:158]: This is an initialization operation. When reading data from the metadata, this list should have been empty. However, due to potential reasons related to gson serialization, the result is null, thus displaying the initialization.

File: fe/fe-core/src/main/java/org/apache/doris/job/manager/JobManager.java [hunk 10]
```diff
@@ -53,37 +54,35 @@ public void start() {
...
     public void registerJob(T job) throws JobException {
         job.checkJobParams();
-        checkJobNameExist(job.getJobName(), job.getJobType(), job.getCurrentDbName());
+        checkJobNameExist(job.getJobName(), job.getJobType());
         if (jobMap.get(job.getJobId()) != null) {
             throw new JobException("job id exist,jobId:" + job.getJobId());
         }
-        //Env.getCurrentEnv().getEditLog().logCreateJob(job);
+        Env.getCurrentEnv().getEditLog().logCreateJob(job);
         //check name exist
```
[the remark this answers is not in the document]
Comment (CalvinKirs) [reply fe/fe-core/src/main/java/org/apache/doris/job/manager/JobManager.java:62]: Oops, I forgot to delete this...

**Comment by zddr** [comment]
run buildall

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by zddr** [comment]
run buildall

**Comment by CalvinKirs** [comment]
run buildall

**Comment by CalvinKirs** [comment]
run P0 Regression (Doris Regression)

**Comment by CalvinKirs** [comment]
run p0

**Comment by CalvinKirs** [comment]
run buildall

**Comment by CalvinKirs** [comment]
run buildall

**Comment by CalvinKirs** [comment]
run buildall

**Comment by CalvinKirs** [comment]
run buildall

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run buildall

**Review by morningman — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVCacheManager.java [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+
+    private static Set<BaseTableInfo> transferTableIfToInfo(List<TableIf> tables) {
+        Set<BaseTableInfo> result = Sets.newHashSet();
+        for (TableIf table : tables) {
+            result.add(new BaseTableInfo(table));
+        }
+        return result;
+    }
+
+    private static Plan getPlanBySql(String querySql, ConnectContext ctx) {
```
Comment (seawinde) [inline fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVCacheManager.java:169]: This plan should be rewrite plan, not physical plan

File: fe/fe-core/src/main/java/org/apache/doris/catalog/MTMV.java [hunk 12]
```diff
@@ -17,16 +17,185 @@
...
+import org.apache.logging.log4j.Logger;
 
 import java.io.DataInput;
+import java.io.DataOutput;
 import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
 
 public class MTMV extends OlapTable {
```
Comment (seawinde) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/MTMV.java:46]: This also can be used by single table, may be materializedView is common?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/MTMV.java [hunk 13]
```diff
@@ -17,16 +17,185 @@
...
+    private MTMVStatus status;
+    @SerializedName("ei")
+    private EnvInfo envInfo;
+    @SerializedName("ji")
+    private MTMVJobInfo jobInfo;
+    @SerializedName("mp")
+    private Map<String, String> mvProperties;
+    @SerializedName("mc")
+    private MTMVCache cache;
+
```
Comment (seawinde) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/MTMV.java:64]: if we add partition mv maintainence, is it extended to add new field easily?

File: fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVCacheManager.java [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+    private boolean containsExternalTable(Set<BaseTableInfo> baseTableInfos) {
+        for (BaseTableInfo baseTableInfo : baseTableInfos) {
+            if (InternalCatalog.INTERNAL_CATALOG_ID != baseTableInfo.getCtlId()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public static MTMVCache generateMTMVCache(MTMV mtmv, ConnectContext ctx) {
```
Comment (seawinde) [inline fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVCacheManager.java:138]: Cache should contain mtmv self

File: fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVCacheManager.java [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+import java.util.Set;
+
+/**
+ * when do some operation, do something about cache
+ */
+public class MTMVCacheManager implements MTMVHookService {
+    private static final Logger LOG = LogManager.getLogger(MTMVCacheManager.class);
+    private Map<BaseTableInfo, Set<BaseTableInfo>> tableMTMVs = Maps.newConcurrentMap();
+
+    public Set<BaseTableInfo> getMtmvsByBaseTable(BaseTableInfo table) {
```
Comment (seawinde) [inline fe/fe-core/src/main/java/org/apache/doris/mtmv/MTMVCacheManager.java:69]: Should also provide list<BaseTableInfo> method param to get materailzed view
and shoud return the materailzed cache

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run feut

**Comment by zddr** [comment]
run buildall

**Comment by zddr** [comment]
run buildall

**Review by seawinde — APPROVED** [review]

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**